scipy/special/_logsumexp.py

```diff
@@ -127,3 +127,88 @@ def logsumexp(a, axis=None, b=None, keepdims=False, return_sign=False):
         return out, sgn
     else:
         return out
+
+
+def softmax(x, axis=None):
+    r"""
+    Softmax function
+
+    The softmax function transforms each element of a collection by
+    computing the exponential of each element divided by the sum of the
+    exponentials of all the elements.  That is, if `x` is a one-dimensional
+    numpy array::
+
+        softmax(x) = np.exp(x)/sum(np.exp(x))
+
+    Parameters
+    ----------
+    x : array_like
+        Input array.
+    axis : int or tuple of ints, optional
+        Axis to compute values along. Default is None and softmax will be
+        computed over the entire array `x`.
+
+    Returns
+    -------
+    s : ndarray
+        An array the same shape as `x`. The result will sum to 1 along the
+        specified axis.
+
+    Notes
+    -----
+    The formula for the softmax function :math:`\sigma(x)` for a vector
+    :math:`x = \{x_0, x_1, ..., x_{n-1}\}` is
+
+    .. math:: \sigma(x)_j = \frac{e^{x_j}}{\sum_k e^{x_k}}
+
+    .. versionadded:: 1.2.0
+
+    Examples
+    --------
+    >>> from scipy.special import softmax
+    >>> np.set_printoptions(precision=5)
+
+    >>> x = np.array([[1, 0.5, 0.2, 3],
+    ...               [1,  -1,   7, 3],
+    ...               [2,  12,  13, 3]])
+    ...
+
+    Compute the softmax transformation over the entire array.
+
+    >>> m = softmax(x)
+    >>> m
+    array([[  4.48309e-06,   2.71913e-06,   2.01438e-06,   3.31258e-05],
+           [  4.48309e-06,   6.06720e-07,   1.80861e-03,   3.31258e-05],
+           [  1.21863e-05,   2.68421e-01,   7.29644e-01,   3.31258e-05]])
+
+    >>> m.sum()
+    1.0000000000000002
+
+    Compute the softmax transformation along the first axis (i.e. the columns).
+
+    >>> m = softmax(x, axis=0)
+
+    >>> m
+    array([[  2.11942e-01,   1.01300e-05,   2.75394e-06,   3.33333e-01],
+           [  2.11942e-01,   2.26030e-06,   2.47262e-03,   3.33333e-01],
+           [  5.76117e-01,   9.99988e-01,   9.97525e-01,   3.33333e-01]])
+
+    >>> m.sum(axis=0)
+    array([ 1.,  1.,  1.,  1.])
+
+    Compute the softmax transformation along the second axis (i.e. the rows).
+
+    >>> m = softmax(x, axis=1)
+    >>> m
+    array([[  1.05877e-01,   6.42177e-02,   4.75736e-02,   7.82332e-01],
+           [  2.42746e-03,   3.28521e-04,   9.79307e-01,   1.79366e-02],
+           [  1.22094e-05,   2.68929e-01,   7.31025e-01,   3.31885e-05]])
+
+    >>> m.sum(axis=1)
+    array([ 1.,  1.,  1.])
+
+    """
+
+    # compute in log space for numerical stability
+    sigma = np.exp(x - logsumexp(x, axis=axis, keepdims=True))
+    return sigma
```

Comment:
Nit by @eric-wieser that I agree wtih: Merge these two lines.

Reply:
I don't have a strong preference here, but since folks are commenting on it a pull request (and I seem to recall seeing it in another one), I'd like to get some clarification.  @hameerabbasi, @eric-wieser, can you give explicit reasons for making this change?  Is it for better performance?  I see about a 5 ns improvement between

    def func1(x):
        y = 2*x
        return y

and

    def func2(x):
        return 2*x

The argument *against* the change is that assigning the return value to a variable, and then returning that variable in a separate `return` statement, eases future degugging.  One can set a break point at the return statement (or add a print statement before it) to inspect the result before it is returned.  Indeed, this pattern is taught in some courses on programming style.

Reply:
Like I said, it's a nit. Feel free to ignore. I tend to only store a variable if it's not a one-liner.

Reply:
Done.

Reply:
The argument was only ever one of brevity, not of performance. Also, it's weird to see:
```
def thing_one():
    thing_two = ...
    return thing_two
```
because it suggests the author wasn't really sure what to call their function.

> to inspect the result before it is returned

Note that pdb provides `__return__` for this purpose